File: Tennis.Play/PlayMatch.cs
using System;
using Tennis.Logic;
using System.Collections.Generic;

namespace Tennis.Play
{
	public class PlayMatch : IPlayMatch
	{
		private readonly IPlaySet playSet;
		private readonly ISide sideOne;
		private readonly ISide sideTwo;
		private Match match;
		private List<SetScore> setScores;
		
		public PlayMatch (IPlaySet playSet, ISide sideOne, ISide sideTwo)
		{
			if (playSet == null)
			{
				throw new ArgumentNullException("playSet");
			}
			
			if (sideOne == null)
			{
				throw new ArgumentNullException("sideOne");
			}
			
			if (sideTwo == null)
			{
				throw new ArgumentNullException("sideTwo");
			}
			
			this.sideOne = sideOne;
			this.sideTwo = sideTwo;
			this.playSet = playSet;
		}
		
		public PlayMatch(ISide sideOne, ISide sideTwo) : this(new PlaySet(sideOne, sideTwo), sideOne, sideTwo)
		{
		}
		
		public ISide Play()
		{
			match = new Match();
			setScores = new List<SetScore>();
			setScores.Add(new SetScore() { Score = match.PrintScore() });
			
			while(match.State != MatchState.MatchWonBySideOne && match.State != MatchState.MatchWonBySideTwo)
			{
				var setWinner = playSet.Play();
				
				match.WinSet(s => setWinner);
				
				setScores.Add(new SetScore() { Score = match.PrintScore(), GameScores = playSet.GetGameScores() });
			}
			
			if (match.State == MatchState.MatchWonBySideOne)
			{
				return sideOne;
			}
			else
			{
				return sideTwo;
			}
		}
		
		public List<SetScore> GetSetScores()
		{
			return setScores;
		}
		
		public ISide SideOne
		{
			get { return this.sideOne; }
		}
		
		public ISide SideTwo
		{
			get { return this.sideTwo; }
		}
	}
}

File: Tennis.FSharp.Play/PlayMatch.fs
namespace Tennis.FSharp.Play
    
    open Tennis.FSharp.Logic
    open FSharpx

    type PlayMatch(playSet: IPlaySet, sideOne: ISide, sideTwo: ISide) =
        let mutable playSet = playSet
        let mutable sideOne = sideOne
        let mutable sideTwo = sideTwo
        let mutable Winner = sideOne
        let tennisMatch = new Match()
        let mutable setsList = []              

        interface IPlayMatch with
            member this.Play() =
                let mutable continueLooping = true
                let setScore = new SetScore()
                setScore.Score <- tennisMatch.PrintScore()
                setsList <- setScore :: setsList
                while continueLooping do    
                    let setWinner = playSet.Play() 
                    let gameWinner = tennisMatch.WinSet(fun x -> setWinner)
                    let activeSetScore = new SetScore()
                    activeSetScore.Score <- tennisMatch.PrintScore()
                    activeSetScore.GameScores <- playSet.GetGameScores()
                    setsList <- activeSetScore :: setsList
                         
                    if tennisMatch.State = MatchState.MatchWonBySideOne then
                        continueLooping <- false
                        Winner <- sideOne
                    elif tennisMatch.State = MatchState.MatchWonBySideTwo then
                        continueLooping <- false
                        Winner <- sideTwo
                Winner

            member this.GetSetScores() =
                List.rev setsList |> List.toSeq

            member this.SideOne
                with get () = sideOne

            member this.SideTwo
                with get () = sideTwo